[WIP] feat: add Cryptopunks snipe strategy inspector

bertmiller opened the pull request:
This PR adds support for classifying snipes of crypto punks. It does not include database integrations.

For background punks are bought and sold in a native marketplace created by LarvaLabs and facilitated by the CryptoPunks contract. To place bids prospective buyers call `enterBidForPunk()`. Given they have bids on their punk, then a seller which wants to sell their punk can call `acceptBidForPunk()`. However, `acceptBidForPunk()` does not tie the sale of a punk to a given bid. It only specifies a _minimum price_ for a punk and accepts the _highest bid_.

The design of the punk bid/acceptances then creates an MEV opportunity: if a bid below market value is going to be accepted then a bot can frontrun the acceptance with a bid of their own slightly outbidding the old bid and sniping the punk away.

### Example snipe
**Example bid acceptance**
https://etherscan.io/tx/0x2330f026de9a3085da664eeba0fd32a1f7be6a8add2ebcc40a49352fcd865ad1
Minimum price: `140000000000000000000`

**Example bid sniping a sale**
https://etherscan.io/tx/0x6237076ae2343a2eea0b90ef8775b95133971963efe0e104310c1d30d4b217e9
Bid: `140000000000000000002`

bertmiller commented:
Another block to test this on would be `13468605`.

There are some other changes that are desired here, like adding tx index, but I wanted to get this PR open early.

gheise commented on mev_inspect/inspect_block.py:
```diff
@@ -91,6 +93,12 @@ async def inspect_block(
+    punk_bid_acceptances = get_punk_bid_acceptances(classified_traces)
+
+    punk_snipes = get_punk_snipes(punk_bids, punk_bid_acceptances)
+    logger.info(f"Block: {block_number} -- Found {len(punk_snipes)} punk snipes")
```
Why is only punk_snipes logged? When len(punk_snipes) == 0 but len(punk_bid_acceptances) > 0, would it not still be useful to still know if any punk transfers occurred in the block?

bertmiller replied:
I was thinking that punk_snipes were the only ones that were MEV relevant. There will probably be many blocks with punk bids but no punk snipes.

gheise commented on mev_inspect/punks.py:
```diff
@@ -0,0 +1,104 @@
+                        from_address=punk_bid.from_address,
+                        punk_index=punk_bid.punk_index,
+                        min_acceptance_price=punk_bid_acceptance.min_price,
+                        acceptance_price=punk_bid.amount,
```
Using price and amount interchangeably introduces unnecessary fuzz. Would be clearer if either only price or only amount were used consistently

bertmiller replied:
Good call out

gheise reviewed (commented):
WAGMI

lukevs commented on mev_inspect/classifiers/specs/cryptopunks.py:
```diff
@@ -0,0 +1,21 @@
+    ClassifierSpec,
+    PunkBidClassifier,
+    PunkBidAcceptanceClassifier,
+)
```
Classifier should be defined here alongside the spec

`mev_inspect.schemas.classifiers` is just for shared schemas

lukevs commented on mev_inspect/schemas/classifiers.py:
```diff
@@ -14,6 +16,28 @@ def get_classification() -> Classification:
+
+    @staticmethod
+    @abstractmethod
+    def get_accept_bid(trace: DecodedCallTrace) -> PunkBidAcceptance:
```
might be a little nicer to move this logic into here: https://github.com/flashbots/mev-inspect-py/pull/138/files?authenticity_token=mYvoUHaaHM4VFXvE7%2BWM7nTzfbV%2Fr%2Fw0jGpRmzeHRcqkMdfn78Ojuv52AFzQO63fltI7QGNHdkEs%2B2itE3r3cg%3D%3D&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension#diff-b7d464d841fd2c44691731a28b6265cb993eeec1f621e2a7fd8526556191a905R59

lukevs replied:
then call the classifier

little bit awkward since there's only one protocol that implements the `PunkBidAcceptance` classifier, but that'd be more consistent

lukevs commented on mev_inspect/schemas/classifiers.py:
```diff
@@ -14,6 +16,28 @@ def get_classification() -> Classification:
+
+    @staticmethod
+    @abstractmethod
+    def get_bid(trace: DecodedCallTrace) -> PunkBid:
```
same comment as for `get_accept_bid`

lukevs commented on mev_inspect/punks.py:
```diff
@@ -0,0 +1,104 @@
+    for punk_bid in punk_bids:
+        for punk_bid_acceptance in punk_bid_acceptances:
+            if punk_bid.punk_index == punk_bid_acceptance.punk_index:
+                if punk_bid.amount > punk_bid_acceptance.min_price:
```
if there's only one bid and it's above the min, that would still generate a "swipe" right?

also we mentioned this creates multiple swipes if there are multiple people who swiped each other (out bid +1 wei then out bid +2 wei, etc)

I think it'd be nicer to have a single object per instance of potentially many bids (`PunkAuction`?)

something like:
```
for punk_bid_acceptance in punk_bid_acceptances:
    punk_bids = get_bids_for_acceptance(...)
    punk_auction = PunkAuction(
        total_bids=len(punk_bids)
        winning_bid_amount=max(bid.amount for bid in punk_bids),
        ...,
    )
```

bertmiller commented on mev:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env bash
```
My man, you were the one that added it

bertmiller commented:
I addressed the above comments. I chose to cycle through each punk bid acceptance and check for the highest bid per punk index from that acceptance. That way there should only be one snipe per bid acceptance.

lukevs commented on mev_inspect/punks.py:
```diff
@@ -0,0 +1,125 @@
+        if highest_punk_bid is None:
+            continue
+
+        if highest_punk_bid.price > punk_bid_acceptance.min_price:
```
if there's only one bid and it happens to be above the min, is that a snipe?
